src/sentry/web/frontend/base.py

```diff
@@ -41,83 +34,108 @@
 
 
 class OrganizationMixin:
+    # This attribute will only be set once determine_active_organization is called.  Subclasses should likely invoke
+    # that method, passing along the organization_slug context that might exist (or might not).
+    active_organization: Optional[ApiOrganization]
+
     # TODO(dcramer): move the implicit organization logic into its own class
     # as it's only used in a single location and over complicates the rest of
     # the code
-    def get_active_organization(self, request: Request, organization_slug=None):
-        """
-        Returns the currently active organization for the request or None
-        if no organization.
+    def determine_active_organization(self, request: Request, organization_slug=None) -> NoReturn:
         """
+        Using the current request and potentially optional organization_slug, 'determines'
+        the current session for this mixin object's scope, placing it into the active_organization attribute.
 
-        # TODO(dcramer): this is a huge hack, and we should refactor this
-        # it is currently needed to handle the is_auth_required check on
-        # OrganizationBase
-        _active_org = getattr(self, "_active_org", None)
-        if _active_org:
-            (active_organization, requesting_user) = _active_org
-            cached_active_org = (
-                active_organization
-                and active_organization.slug == organization_slug
-                and requesting_user == request.user
-            )
-            if cached_active_org:
-                return active_organization
-
-        active_organization = None
 
-        is_implicit = organization_slug is None
+        Note that this function has some side effects and should only be called once at the 'head' of a request.
+        """
+        # TODO: Really, this should be a service object, not a mixin, but unfortunately extracting that is not trivial
+        # due to the depth of references to this logic.
 
-        if is_implicit:
-            organization_slug = request.session.get("activeorg")
-            if request.subdomain is not None and request.subdomain != organization_slug:
-                # Customer domain is being used, set the subdomain as the requesting org slug.
-                organization_slug = request.subdomain
+        assert not hasattr(
```

Comment:
Now that I removed the caching "hack" I wanted tests to verify the assumption that this side-effect laden function is only invoked once per request.